### DISCUSSION TITLE
Improved Magma Documents

### DISCUSSION THREAD

**Discussion by hcgatewood** [magma/magma#5587] [Ideas]
# Improved Magma Documents

This proposal outlines a way to support Magma's document needs in its transition to fully open-source development and governance. *Documents* encompasses shared communication which don't fit neatly onto GitHub or Slack.

## tl;dr

- Magma-specific workspaces: Google Drive, Quip, Zoom
- Skip CI checks for new `internal/` directory for internal documentation
- [SOPS](https://github.com/mozilla/sops) for secrets management

## Context

There are 3 categories of documents, in the broadest consideration

- Documentation (public and internal)
- Traditional documents
- Secrets

Documentation is generally readmes, with specific focus on the readmes that constitute Magma's Docusaurus site. Documents include non-readme assets such as slides, multimedia, and transient information. Secrets include CI secrets, and any future need for project-wide secrets.

### Documentation: `internal/` directory

The CI team has expressed the desire for a collaboration mechanism that doesn't pollute the `magma/magma` repo. I think it's worthwhile to keep to the monorepo pattern, so, instead, we can use the top-level `internal/` directory as a location for public documents which are not relevant to the greater community. We can explicitly disable CI checks for this directory, which should allow low-friction updates.

### Traditional documents: shared docs suites

Magma needs a shared docs suite. That is, we can't collaborate entirely through GitHub -- consider the following

- Slides/presentations
- Multimedia: images, videos
- Transient information
    - Docs with no long-term importance
    - Docs requiring fast iteration, or where only final output is important
- High-input design docs (PRs aren't ergonomic enough when comment volume exceeds 10-20 comments)

While it's conceivable that we use an open-source docs suite and check everything into version control, this feels unnecessarily painful and will inevitably be bypassed by suboptimal workarounds.

The immediate solution is a Magma-specific *Google Drive*, opting to be highly permissive in adding new members to the team.

A Magma-specific *Quip* is also desirable, as Quip is written in, and directly exports to, Markdown, which is ideal for technical documentation. An alternative solution here might be a [StackEdit](https://github.com/benweet/stackedit) deployment backed by the Magma Google Drive.

Finally, a Magma-specific *Zoom* team would remove our current reliance on OIF's Zoom team.

### Secrets: SOPS

Watch this video for an [intro to SOPS](https://www.youtube.com/watch?v=V2PRhxphH2w&ab_channel=SecuringDevOps).

SOPS would allow us to store the project's secrets directly in a top-level `secrets/` directory, e.g. `secrets/ci.yml`. The root PGP key can be held in trust by the TSC (e.g. in a Magma-specific Lastpass account, or comparable alternative).

A non-root PGP key can also be added for use in CI. This would allow a generic solution for retrieving CI secrets -- include the non-root PGP key as a CI secret, then use that to decrypt the full set of additional CI secrets needed from their encrypted form in the `secrets/` directory.

A yearly GitHub issue can be automatically opened to rotate the full set of keys.

**Comment by jeremy-donson** [comment]
+1 for sops 

We have a few needs for docs:
- Docusaurus edits ought to only kick off docs-related ci jobs => This is on my radar now.
- Review of those public doc edits might be best done with a noob?
- It is the non-public dev ops support docs and tools that I am most concerned about.

Glad to share them with the magma core.org team, and I would expect them to be GitHub REAME.md files.
This does assume that all magma team members have a magmacore.org email address. => See Phil Ritter for one.
We can opt to share that info selectively.  This makes them easy to promote to public in docusaurus.

I also have some DevOps doc tags to suggest, but clearly discussions are NOT the place for that.
My guess is that we can escalate discussions by adding them to tagged hub issues or PR's.
Alternatively, maybe zen hub could be used for this?

**Reply by hcgatewood** [reply]
I'm not categorically opposed to having a separate repo -- either here or in DM, can you give me some concrete examples of docs that would better fit in an internal repo?

**Comment by karthiksubraveti** [comment]
One more think I can think of why this is useful, this can be a good place to share more information on release testing(cc @sudhikan ) with the wider community.

**Comment by electronjoe** [comment]
Minor nit! I think this proposal hoovers up a bit more than necessary, as I find myself wanting to comment on sub-parts of it (or in some cases, hesitating to generally thumbs up or down).  Like a large PR =D Would ideally be like 2 or 3 proposals.

- SOPS sounds good, would love to see a prototype before we commit?
- On `/internal/`, can we clarify how this is different than /docs today?
  - Perhaps internal is a common naming? Does it change what's within (rel `/docs/`) or..
- On the Magma-specific workspaces: Google Drive, Quip, Zoom
  - Zoom: Ability to host meetings seems important
  - Google Drive: Mixed feelings
    - Understand the need
    - Not a fan of this solution but maybe nothing better exists..
    - Creating yet another account for everybody involved
    - Ephemeral nature of docs that I think **shouldn't be** ephemeral
    - Lures people into design docs on Google Docs (I consider an anti-pattern)
    - Access permissions annoyances for the public
    - For the sake of discussion, why not export PDFs and post to Dropbox or some such? Or Magma blog?
  - Quip
    - Seems a useful workflow mechanism for proposal authorship
    - This or some other Markdown-editor-with-commenting seems helpful

**Reply by hcgatewood** [reply]
SOPS: I think for now our setup is "good enough", but we can keep this as the plan for when we hit scale issues

Internal: similar idea as^

Workspaces: I understand the hesitance for GDrive, but I really don't view checking end-assets into VCS as a realistic solution. The affordances of a shared workspace, where we have access to the intermediary artifacts of the end-asset (slides, images, comment history, etc) has been super valuable for the FB-internal Magma development, and I don't see that value dissipating at all as we move to fully open-sourced development pattern
